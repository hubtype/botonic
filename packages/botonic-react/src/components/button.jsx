import { INPUT, params2queryString } from '@botonic/core'
import React, { useContext, useState } from 'react'
import styled from 'styled-components'

import { COLORS, WEBCHAT } from '../constants'
import { WebchatContext } from '../contexts'
import { renderComponent } from '../util/react'
import { ButtonsDisabler } from './buttons-disabler'

const StyledButton = styled.button`
  display: flex;
  width: 100%;
  max-height: 80px;
  font-size: 14px;
  text-align: center;
  align-content: center;
  justify-content: center;
  padding: 12px 32px;
  font-family: inherit;
  border: none;
  border: 1px solid ${COLORS.SEASHELL_WHITE};
  cursor: pointer;
  outline: 0;
  border-top-right-radius: ${props => props.top || '0px'};
  border-top-left-radius: ${props => props.top || '0px'};
  border-bottom-right-radius: ${props => props.bottom || '0px'};
  border-bottom-left-radius: ${props => props.bottom || '0px'};
  overflow: hidden;
`

export const Button = props => {
  const {
    webchatState,
    openWebview,
    sendPayload,
    sendInput,
    getThemeProperty,
    updateMessage,
  } = useContext(WebchatContext)
  const [hover, setHover] = useState(false)
  const { theme } = webchatState
  const { autoDisable, disabledStyle } = ButtonsDisabler.resolveDisabling(
    webchatState.theme,
    props
  )
  const handleClick = event => {
    event.preventDefault()
    const type = getThemeProperty(
      WEBCHAT.CUSTOM_PROPERTIES.buttonMessageType,
      INPUT.TEXT
    )
    if (props.webview) openWebview(props.webview, props.params)
    else if (props.path) {
      type == INPUT.POSTBACK
        ? sendPayload(`__PATH_PAYLOAD__${props.path}`)
        : sendInput({
            type: INPUT.TEXT,
            data: String(props.children),
            payload: `__PATH_PAYLOAD__${props.path}`,
          })
    } else if (props.payload) {
      type == INPUT.POSTBACK
        ? sendPayload(props.payload)
        : sendInput({
            type: INPUT.TEXT,
            data: String(props.children),
            payload: props.payload,
          })
    } else if (props.url) {
      window.open(props.url, props.target || '_blank')
    }
    if (props.onClick) props.onClick()
    if (props.setDisabled) {
      props.setDisabled(true)
      const messageToUpdate = webchatState.messagesJSON.filter(
        m => m.id == props.parentId
      )[0]
      const updatedMsg = ButtonsDisabler.getUpdatedMessage(messageToUpdate, {
        autoDisable,
        disabledStyle,
      })
      updateMessage(updatedMsg)
    }
  }

  const renderBrowser = () => {
    const buttonStyle = getThemeProperty(WEBCHAT.CUSTOM_PROPERTIES.buttonStyle)
    const CustomButton = getThemeProperty(
      WEBCHAT.CUSTOM_PROPERTIES.customButton
    )
    if (CustomButton) {
      return (
        <div onClick={e => handleClick(e)}>
          <CustomButton>{props.children}</CustomButton>
        </div>
      )
    }

    const buttonBgColor = hover
      ? getThemeProperty(
          WEBCHAT.CUSTOM_PROPERTIES.buttonHoverBackground,
          COLORS.CONCRETE_WHITE
        )
      : getThemeProperty(
          WEBCHAT.CUSTOM_PROPERTIES.buttonStyleBackground,
          COLORS.SOLID_WHITE
        )
    const buttonTextColor = hover
      ? getThemeProperty(
          WEBCHAT.CUSTOM_PROPERTIES.buttonHoverTextColor,
          COLORS.SOLID_BLACK
        )
      : getThemeProperty(
          WEBCHAT.CUSTOM_PROPERTIES.buttonStyleColor,
          COLORS.SOLID_BLACK
        )

    return (
      <StyledButton
        theme={theme}
        onMouseEnter={() => setHover(true)}
        onMouseLeave={() => setHover(false)}
        onClick={e => handleClick(e)}
        style={{
          ...buttonStyle,
          color: buttonTextColor,
          backgroundColor: buttonBgColor,
          ...(props.disabled && autoDisable && disabledStyle),
        }}
        bottom={props.bottomRadius}
      >
        {props.children}
      </StyledButton>
    )
  }

  const renderNode = () => {
    const disabledProps = ButtonsDisabler.constructNodeProps(props)
    if (props.webview) {
      const Webview = props.webview
      let params = ''
      if (props.params) params = params2queryString(props.params)
      return (
        <button url={`/webviews/${Webview.name}?${params}`} {...disabledProps}>
          {props.children}
        </button>
      )
    } else if (props.path) {
      const payload = `__PATH_PAYLOAD__${props.path}`
      return (
        <button payload={payload} {...disabledProps}>
          {props.children}
        </button>
      )
    } else if (props.payload) {
      return (
        <button payload={props.payload} {...disabledProps}>
          {props.children}
        </button>
      )
    } else if (props.url) {
      return (
        <button url={props.url} target={props.target} {...disabledProps}>
          {props.children}
        </button>
      )
    } else if (props.onClick) {
      return null
    }
    throw new Error('Button missing payload, path, webviews, url or onClick')
  }

  return renderComponent({ renderBrowser, renderNode })
}

Button.serialize = buttonProps => {
  let payload = buttonProps.payload
  if (buttonProps.path) payload = `__PATH_PAYLOAD__${buttonProps.path}`
  return {
    button: {
      payload,
      url: buttonProps.url,
      target: buttonProps.target,
      webview: buttonProps.webview && String(buttonProps.webview),
      title: buttonProps.children && String(buttonProps.children),
      ...ButtonsDisabler.withDisabledProps(buttonProps),
    },
  }
}
