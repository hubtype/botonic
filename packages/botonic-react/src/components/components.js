/**
 * Get the name of an element (instantiated component)
 * @param {React.Element} component
 * @returns {string}
 */
export function getElementName(component) {
  return component.type.name
}

/**
 * Get the name of a component TYPE
 * @param {React.ComponentType} component
 * @returns {string}
 */
export function getComponentTypeName(component) {
  return component.displayName || component.name || 'Component'
}
