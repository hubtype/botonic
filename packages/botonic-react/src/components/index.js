export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Image } from './image'
export { Location } from './location'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export { WebchatSettings } from './webchat-settings'
export { WhatsappTemplate } from './whatsapp-template'
