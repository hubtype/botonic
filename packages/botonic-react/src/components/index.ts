export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Image } from './image'
export * from './index-types'
export { Location } from './location'
export { Message } from './message'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Raw } from './raw'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export { WebchatSettings } from './webchat-settings'
export {
  WhatsappButtonList,
  WhatsappButtonListProps,
  WhatsappButtonListRowProps,
  WhatsappButtonListSectionProps,
} from './whatsapp-button-list'
export {
  WhatsappCTAUrlButton,
  WhatsappCTAUrlButtonProps,
} from './whatsapp-cta-url-button'
export { WhatsappTemplate } from './whatsapp-template'
