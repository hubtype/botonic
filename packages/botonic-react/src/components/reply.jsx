import React, { useContext } from 'react'
import styled from 'styled-components'

import { COLORS, WEBCHAT } from '../constants'
import { WebchatContext } from '../contexts'
import { renderComponent } from '../util/react'

const StyledButton = styled.button`
  width: 100%;
  padding: 4px 8px;
  font-family: inherit;
  border-radius: 8px;
  cursor: pointer;
  outline: 0;
`

export const Reply = props => {
  const { sendText, getThemeProperty } = useContext(WebchatContext)
  const handleClick = event => {
    event.preventDefault()
    if (props.children) {
      let payload = props.payload
      if (props.path) payload = `__PATH_PAYLOAD__${props.path}`
      sendText(props.children, payload)
    }
  }

  const renderBrowser = () => {
    const replyStyle = getThemeProperty(WEBCHAT.CUSTOM_PROPERTIES.replyStyle)
    const CustomReply = getThemeProperty(WEBCHAT.CUSTOM_PROPERTIES.customReply)
    if (CustomReply) {
      return (
        <div onClick={e => handleClick(e)}>
          <CustomReply>{props.children}</CustomReply>
        </div>
      )
    }

    return (
      <StyledButton
        style={{
          border: `1px solid ${getThemeProperty(
            WEBCHAT.CUSTOM_PROPERTIES.brandColor,
            COLORS.BOTONIC_BLUE
          )}`,
          color: getThemeProperty(
            WEBCHAT.CUSTOM_PROPERTIES.brandColor,
            COLORS.BOTONIC_BLUE
          ),
          ...replyStyle,
        }}
        onClick={e => handleClick(e)}
      >
        {props.children}
      </StyledButton>
    )
  }

  const renderNode = () => {
    if (props.path) {
      const payload = `__PATH_PAYLOAD__${props.path}`
      return <reply payload={payload}>{props.children}</reply>
    }
    return <reply payload={props.payload}>{props.children}</reply>
  }

  return renderComponent({ renderBrowser, renderNode })
}

Reply.serialize = replyProps => {
  let payload = replyProps.payload
  if (replyProps.path) payload = `__PATH_PAYLOAD__${replyProps.path}`
  return { reply: { title: replyProps.children, payload } }
}
