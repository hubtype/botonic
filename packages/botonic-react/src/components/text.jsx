import React from 'react'
import { Message } from './message'

import { INPUT } from '@botonic/core'

import { serializeMarkdown, toMarkdownChildren } from './markdown'

const serializeText = children => {
  children = Array.isArray(children) ? children : [children]
  const text = children
    .filter(e => !e.type)
    .map(e => {
      if (Array.isArray(e)) return serializeText(e)
      else return String(e)
    })
    .join('')
  return text
}

const serialize = textProps => {
  if (!textProps.markdown)
    return {
      text: serializeText(textProps.children),
    }
  return { text: serializeMarkdown(textProps.children) }
}

export const Text = props => {
  const defaultTextProps = {
    markdown: props.markdown === undefined ? true : props.markdown,
  }
  const textProps = {
    ...props,
    ...defaultTextProps,
    ...{ children: React.Children.toArray(props.children) },
  }
  if (!textProps.markdown)
    return (
      <Message json={serialize(textProps)} {...textProps} type={INPUT.TEXT}>
        {textProps.children}
      </Message>
    )
  return (
    <Message json={serialize(textProps)} {...textProps} type={INPUT.TEXT}>
      {toMarkdownChildren(textProps.children)}
    </Message>
  )
}

Text.serialize = serialize
