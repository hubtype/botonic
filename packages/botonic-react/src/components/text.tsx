import { INPUT } from '@botonic/core'
import React, { Children } from 'react'

import { mapObjectNonBooleanValues } from '../util/react'
import { TextProps } from './index-types'
import { serializeMarkdown, toMarkdownChildren } from './markdown'
import { Message } from './message'

const serializeText = children => {
  children = Array.isArray(children) ? children : [children]
  const text = children
    .filter(e => !e.type)
    .map(e => {
      if (Array.isArray(e)) return serializeText(e)
      else return String(e)
    })
    .join('')
  return text
}

const serialize = (textProps: TextProps) => {
  if (!textProps.markdown)
    return {
      text: serializeText(textProps.children),
    }
  return { text: serializeMarkdown(textProps.children) }
}

export const Text = (props: TextProps) => {
  const defaultTextProps = {
    markdown: props.markdown === undefined ? true : props.markdown,
    feedbackenabled: props.feedbackenabled,
    inferenceid: props.inferenceid,
  }

  const textProps = mapObjectNonBooleanValues({
    ...props,
    ...defaultTextProps,
    ...{ children: Children.toArray(props.children) },
  })

  if (!textProps.markdown)
    return (
      <Message json={serialize(textProps)} {...textProps} type={INPUT.TEXT}>
        {textProps.children}
      </Message>
    )
  return (
    <Message json={serialize(textProps)} {...textProps} type={INPUT.TEXT}>
      {toMarkdownChildren(textProps.children)}
    </Message>
  )
}

Text.serialize = serialize
