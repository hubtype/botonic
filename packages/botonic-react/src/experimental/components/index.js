export { Button } from '../../components/button'
export { Element } from '../../components/element'
export { MessageTemplate } from '../../components/message-template'
export * from '../../components/multichannel'
export { Pic } from '../../components/pic'
export { Reply } from '../../components/reply'
export { ShareButton } from '../../components/share-button'
export { Subtitle } from '../../components/subtitle'
export { Title } from '../../components/title'
export { WebchatSettings } from '../../components/webchat-settings'

// Experimental
export { Audio } from './audio'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Image } from './image'
export { Location } from './location'
export { Text } from './text'
export { Video } from './video'
export { WhatsappTemplate } from './whatsapp-template'
