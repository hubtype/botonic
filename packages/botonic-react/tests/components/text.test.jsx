import React from 'react'
import { Text } from '../../src/components/text'
import { Button } from '../../src/components/button'
import { Reply } from '../../src/components/reply'
import TestRenderer from 'react-test-renderer'

const renderToJSON = sut => TestRenderer.create(sut).toJSON()

describe('Text Component', () => {
  test('Just one Text', () => {
    const sut = <Text>Just one Text</Text>
    const tree = renderToJSON(sut)
    expect(tree).toMatchSnapshot()
  })

  test('Text with 1 button ', () => {
    const sut = (
      <Text>
        Text with 1 button
        <Button key={'1'} payload='payload1'>
          Text 1
        </Button>
      </Text>
    )
    const tree = renderToJSON(sut)
    expect(tree).toMatchSnapshot()
  })

  test('Text with N button ', () => {
    const sut = (
      <Text>
        Text with N button
        {[
          <Button key={'1'} payload='payload1'>
            Text 1
          </Button>,
          <Button key={'2'} payload='payload2'>
            Text 2
          </Button>,
        ]}
      </Text>
    )
    const tree = renderToJSON(sut)
    expect(tree).toMatchSnapshot()
  })

  test('N Text with N button ', () => {
    const sut = (
      <>
        {[1, 2].map((e, i) => {
          return (
            <Text key={i}>
              N Text with N button {i + 1}
              <Button payload={`payload${i + 1}`}>Button Text {i + 1}</Button>
            </Text>
          )
        })}
      </>
    )
    const tree = renderToJSON(sut)
    expect(tree).toMatchSnapshot()
  })
  test('Text with 1 reply ', () => {
    const sut = (
      <Text>
        Text with 1 reply
        <Reply key={'1'} payload='payload1'>
          Text Reply 1
        </Reply>
      </Text>
    )
    const tree = renderToJSON(sut)
    expect(tree).toMatchSnapshot()
  })

  test('Text with N replies ', () => {
    const sut = (
      <Text>
        Text with N replies
        {[
          <Reply key={'1'} payload='payload1'>
            Text Reply 1
          </Reply>,
          <Reply key={'2'} payload='payload2'>
            Text Reply 2
          </Reply>,
        ]}
      </Text>
    )
    const tree = renderToJSON(sut)
    expect(tree).toMatchSnapshot()
  })

  test('N Text with N replies ', () => {
    const sut = (
      <>
        {[1, 2].map((e, i) => {
          return (
            <Text key={i}>
              N Text with N replies {i + 1}
              <Reply payload={`payload${i + 1}`}>Reply Text {i + 1}</Reply>
            </Text>
          )
        })}
      </>
    )
    const tree = renderToJSON(sut)
    expect(tree).toMatchSnapshot()
  })
})
