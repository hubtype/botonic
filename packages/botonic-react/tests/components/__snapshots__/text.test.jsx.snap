// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component Just one Text 1`] = `
<message
  delay={0}
  markdown={true}
  type="text"
  typing={0}
>
  Just one Text
</message>
`;

exports[`Text Component N Text with N button  1`] = `
Array [
  <message
    delay={0}
    markdown={true}
    type="text"
    typing={0}
  >
    N Text with N button 
    1
    <button
      payload="payload1"
    >
      Button Text 
      1
    </button>
  </message>,
  <message
    delay={0}
    markdown={true}
    type="text"
    typing={0}
  >
    N Text with N button 
    2
    <button
      payload="payload2"
    >
      Button Text 
      2
    </button>
  </message>,
]
`;

exports[`Text Component N Text with N replies  1`] = `
Array [
  <message
    delay={0}
    markdown={true}
    type="text"
    typing={0}
  >
    N Text with N replies 
    1
    <reply
      payload="payload1"
    >
      Reply Text 
      1
    </reply>
  </message>,
  <message
    delay={0}
    markdown={true}
    type="text"
    typing={0}
  >
    N Text with N replies 
    2
    <reply
      payload="payload2"
    >
      Reply Text 
      2
    </reply>
  </message>,
]
`;

exports[`Text Component Text with 1 button  1`] = `
<message
  delay={0}
  markdown={true}
  type="text"
  typing={0}
>
  Text with 1 button
  <button
    payload="payload1"
  >
    Text 1
  </button>
</message>
`;

exports[`Text Component Text with 1 reply  1`] = `
<message
  delay={0}
  markdown={true}
  type="text"
  typing={0}
>
  Text with 1 reply
  <reply
    payload="payload1"
  >
    Text Reply 1
  </reply>
</message>
`;

exports[`Text Component Text with N button  1`] = `
<message
  delay={0}
  markdown={true}
  type="text"
  typing={0}
>
  Text with N button
  <button
    payload="payload1"
  >
    Text 1
  </button>
  <button
    payload="payload2"
  >
    Text 2
  </button>
</message>
`;

exports[`Text Component Text with N replies  1`] = `
<message
  delay={0}
  markdown={true}
  type="text"
  typing={0}
>
  Text with N replies
  <reply
    payload="payload1"
  >
    Text Reply 1
  </reply>
  <reply
    payload="payload2"
  >
    Text Reply 2
  </reply>
</message>
`;
